### DISCUSSION TITLE
xpath bug？

### DISCUSSION THREAD

**Discussion by kloselhy** [zeux/pugixml#421] [Q&A] [answered]
A xml like this:
### <root>
    <item>
        <a>first a</a>
        <a>second a</a>
    </item>
    <item>
        <a>third a</a>
        <a>fourth a</a>
    </item>
</root>

why "//a[1]" return {"first a","third a"} and "//a[2]" return {"second a","fourth a"}?

I think the correct answer should be like this：
"//a[1]" return "first a"；
"//a[2]" return "second a"；
"//a[3]" return "third a"；
"//a[4]" return "fourth a"；

is anything wrong in the code?

**Comment by zeux** [comment, marked as answer]
This is how XPath works. There's a difference between `(//a)[2]` and `//a[2]` - the former returns the second element of a set returned by `//a`; the latter returns the second `a` child for each level in the hierarchy that `//` traverses.

See the following note from XPath spec (https://www.w3.org/TR/1999/REC-xpath-19991116/):

> NOTE: The location path //para[1] does not mean the same as the location path /descendant::para[1]. The latter selects the first descendant para element; the former selects all descendant para elements that are the first para children of their parents.

**Reply by kloselhy** [reply]
Thanks for your reply. I've modified my program and it works fine now.
It really bothered me before, because the returned nodeset is different from libxml returns, now the problem is solved.
Thanks again for your help.
